## Overview

This pacakge provides an interface to estimate transcript abundances of any samples quantified by the aligner *rail*. This method is a non-negative least squares (NNLS) estimation that calculates the number of reads that originated from each transcript of the coding portion of the GencodeV25 transcriptome. The model does not require raw aligned `bam` files, but is content with compressed coverage statistics primarily stored in `bigwig` formats. 

## Installation
Prior to the installation of `recountNNLS` please install the companion data package `recountNNLSdata`:

```{r, eval=F}
source("https://bioconductor.org/biocLite.R")
dev_tools::install_github("recountNNLSdata")
dev_tools::biocLite("recountNNLS", build_vignettes=TRUE)
```

## Quantifying Samples on recount2

To use our NNLS model on the samples on recount2, all one needs is the SRA project id.

```{r, eval=F}
library(recountNNLS)

## Specify a SRA project and download the relevant path data from recount2
project = 'SRP063581'
pheno = processPheno(project)

## Compile the coverage of the relevant exonic segments
counts_ex = getExCounts(pheno)

## Compile the coverage of the relevant junctions
counts_jx = getJxCounts(project)

## Main NNLS workhorse function to create a RSE of transcript abundance
rse_tx = recountNNLS(pheno, counts_ex, counts_jx)

```
## Data Not Yet Part of recount2

Please see vignette `vignette('recountNNLS', package='recountNNLS')`.

## Deliverables

The output of `recountNNLS()` is a `RangeSummarizedExperiment` where:

* Each row corresponds to a transcript
* Each column corresponds to a run (sample)
* `rowRanges()` can be used to access a `GRangesList` annotation of the transcripts
* `assays()` returns a list of length 2 that can be access with `$`. `$counts` returns a matrix of estimated transcript abundance on a number of reads scale. Using `$se` returns a matrix of estimated standard error of the abundance



